libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -2422,30 +2437,30 @@ instance (Ord a, HasSpec fn a) => Semigroup (SetSpec fn a) where
 instance (Ord a, HasSpec fn a) => Monoid (SetSpec fn a) where
   mempty = SetSpec mempty mempty TrueSpec
 
+-- ==================================================
+-- THIS FUNCTION IS ONLY USED ONCE in  genFromTypeSpec  for SETS
+-- I think it might be a first attempt at cardinality.
+-- MAYBE WE CAN REPLACE IT with Cardinality
+-- ===================================================
 -- NOTE: No guarantees that this is a tight spec!
-sizeSpec :: forall fn a. (Ord a, BaseUniverse fn) => Spec fn (Set a) -> Spec fn Int
+sizeSpec :: forall fn a. (Ord a, BaseUniverse fn, HasSpec fn a) => Spec fn (Set a) -> Spec fn Size
 sizeSpec TrueSpec = TrueSpec
-sizeSpec (MemberSpec es) = MemberSpec (Set.size <$> es)
+sizeSpec (MemberSpec es) = MemberSpec (setSize <$> es)
 sizeSpec SuspendedSpec {} = TrueSpec
 sizeSpec (ErrorSpec es) = ErrorSpec es
-sizeSpec (TypeSpec (SetSpec must elemSpec sizeSpec) cant) =
+sizeSpec (TypeSpec (SetSpec must elemSpec szSpec) cant) =
   cantSpec
-    <> sizeSpec
-    <> geqSpec @fn (Set.size must)
-    <> specSizeBound elemSpec
+    <> szSpec
+    <> geqSpec @fn (setSize must)
+    <> specIntToSize (cardinality elemSpec)
   where
+    -- I don't see how this can have any effect on the sizeSpec
+    -- Please help me understand.
+
```

Comment:
Max, is this a first version of cardinality? Help me understand line 2455

Reply:
Yes, there is something wrong here. It should be `atMostSpec (cardinality elemSpec)` for some `atMostSpec :: Spec fn Int -> Spec fn Int`.